### PR TITLE
Enable incremental compaction on off-strategy

### PR DESCRIPTION
**Pull request by raphaelsc**
Off-strategy suffers with a 100% space overhead, as it adopted
a sort of all or nothing approach. Meaning all input sstables,
living in maintenance set, are kept alive until they're all
reshaped according to the strategy criteria.

Input sstables in off-strategy are very likely to be mostly disjoint,
so it can greatly benefit from incremental compaction.

The incremental compaction approach is not only good for
decreasing disk usage, but also memory usage (as metadata of
input and output live in memory), and file desc count, which
takes memory away from OS.

Turns out that this approach also greatly simplifies the
off-strategy impl in compaction manager, as it no longer have
to maintain new unused sstables and mark them for
deletion on failure, and also unlink intermediary sstables
used between reshape rounds.

Fixes https://github.com/scylladb/scylladb/issues/14992.

### CODE REVIEW

**Comment by Annamikhlin** [comment]
https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3659/ - job was aborted due to issue with `spider2` (local server)

File: compaction/compaction_manager.cc [hunk 1]
```diff
@@ -1358,33 +1344,20 @@ class offstrategy_compaction_task_executor : public compaction_task_executor, pu
...
-        };
-        co_await _cm.on_compaction_completion(t, std::move(completion_desc), sstables::offstrategy::yes);
+        // There might be some remaining sstables in maintenance set that didn't require reshape, or the
+        // user has aborted off-strategy. So we can only integrate them into the main set, such that
+        // they become candidates for regular compaction. We cannot hold them forever in maintenance set,
+        // as that causes read and space amplification issues.
+        if (auto sstables = get_reshape_candidates(); sstables.size()) {
+            auto completion_desc = sstables::compaction_completion_desc{
+                .old_sstables = sstables, // removes from maintenance set.
+                .new_sstables = sstables, // adds into main set.
```
Comment (denesb) [inline compaction/compaction_manager.cc:1356]: I don't understand: we cannot add staging sstables into the main set until view updates were not generated for them. Or is off-strategy triggered after that already happened?

Comment (raphaelsc) [reply compaction/compaction_manager.cc:1356]: get_reshape_candidates() filters out candidates requiring view building (i.e. they sit on staging). we trigger off-strategy timer when moving sstables from staging, but it can happen that new sstables are added to staging while off-strategy is running (and those will be filtered out).

Comment (denesb) [reply compaction/compaction_manager.cc:1356]: I see.

**Comment by denesb** [comment]
LGTM, we just need a passing CI to queue.

File: compaction/compaction.cc [hunk 2]
```diff
@@ -729,6 +731,11 @@ class compaction {
 
         _ms_metadata.min_timestamp = timestamp_tracker.min();
         _ms_metadata.max_timestamp = timestamp_tracker.max();
+
+        size_t max_overlapping = std::ranges::max(token_map | boost::adaptors::transformed([] (auto& interval) {
+            return interval.second.size();
+        }));
```
Comment (avikivity) [inline compaction/compaction.cc:737]: std::ranges::max supports projection, so you should be able to write

```c++
std::ranges::max(token_map, std::less<>(), [] (const interval_type& interval) { return interval.second.size(); });
```

File: compaction/compaction.cc [hunk 3]
```diff
@@ -729,6 +731,11 @@ class compaction {
...
+        _incremental_compaction_worthwhile = max_overlapping < _sstables.size();
```
Comment (avikivity) [inline compaction/compaction.cc:738]: I think this generates some false negatives. If we have one run with two sstables S1, S2, and another run with one sstable S3, we get four elements in token_map: one for S1 only, one for S1+S3, one for S2+S3, one for S2. But it should definitely be enabled for incremental compaction.

Comment (avikivity) [reply compaction/compaction.cc:738]: Sorry, I was misinterpreting the computation. But why do you check against < sstable.size()? Suppose you have one non-overlapping sstable and 200 overlapping sstables. You'd get max_overlapping=200 and sstables.size() = 201.

Comment (raphaelsc) [reply compaction/compaction.cc:738]: then (max_overlapping == 2) < (_sstables.size() == 3). incremental is worth being enabled if we detect that at any point in time there will be less sstables opened than there are in total, and that's what I implemented.

File: compaction/compaction.cc (hunk 2 above)
Comment (raphaelsc) [reply compaction/compaction.cc:737]: will change, thanks.

File: compaction/compaction.cc (hunk 3 above)
Comment (raphaelsc) [reply compaction/compaction.cc:738]: > Sorry, I was misinterpreting the computation. But why do you check against < sstable.size()? Suppose you have one non-overlapping sstable and 200 overlapping sstables. You'd get max_overlapping=200 and sstables.size() = 201.

That's a good counter example. So I can change it to enable incremental if there's any non-overlapping sstable.

Comment (raphaelsc) [reply compaction/compaction.cc:738]: Actually, that's wrong, as there could be overlapping among all intervals, but it's still worth incremental.

Comment (raphaelsc) [reply compaction/compaction.cc:738]: > Sorry, I was misinterpreting the computation. But why do you check against < sstable.size()? Suppose you have one non-overlapping sstable and 200 overlapping sstables. You'd get max_overlapping=200 and sstables.size() = 201.

Wait, that would still be correct. 200 < 201, so incremental would be enabled, and that's correct because the disjoint sstable (or the overlapping ssts; depending on their token order) can be released earlier. False positive is ok (just some additional logic seeing nothing can be released), false negative is not.

Comment (avikivity) [reply compaction/compaction.cc:738]: Then why not always be incremental? Is there a downside?

I guess generation of the gc sstable is a downside.

But I don't trust the calculation. How do we know it doesn't generate more false positives, or false negatives?

Comment (raphaelsc) [reply compaction/compaction.cc:738]: We still only enable incremental compaction for ICS and LCS. We check max allowed size for output fragments too. So STCS and TWCS are unaffected by this change.

Effect of today's _contains_multi_fragment_runs is that any level above 0 will enable incremental compaction.

On "level 0", max overlapping == # of input runs.

Even on levels greater than 0, where memtables are way smaller than 0, max overlapping is still == # of runs.

That's because compaction output a single-fragment run if the output size is < 1G.

With this new calculation, that's preserved.

max overlapping will only be equal to # of input sstables on levels that contain single-fragment runs. And in practice, level 0 sstables all overlap one another at some token T.
In other words, "level 0" compaction's max overlapping will equal the number of input sstables.
level 1 and above, containing multi-fragment runs, will have less overlapping than the number of input sstables.

So the semantics is still the same, but we no longer rely on run id to determine activation of incremental.

Comment (avikivity) [reply compaction/compaction.cc:738]: ok, makes sense.

So what does the test check? that the maximum overlap is less than the number of sstables? It's a really low bar.

Comment (avikivity) [reply compaction/compaction.cc:738]: So why not it unconditionally for LCS/ICS?

Comment (raphaelsc) [reply compaction/compaction.cc:738]: single large table == single table taking more than 50% of space.

**Comment by raphaelsc** [comment]
v2
- Rebased
- Fixed std::ranges::max() over empty range (UB)
- Use std::views::transform instead of boost variant.
- Preserved behavior where L0 compactions don't have incremental compaction enabled on them, to avoid bad surprises when backporting this series.

**Comment by raphaelsc** [comment]
v3
don't try to replace sstables in reshape_compaction when the context is not off-strategy (no replacer method is feeded).

**Comment by raphaelsc** [comment]
@denesb please consider merging it.

**Comment by denesb** [comment]
> @denesb please consider merging it.

I want an ack from @avikivity.

File: compaction/compaction.cc (hunk 3 above)
Comment (avikivity) [reply compaction/compaction.cc:743]: I mean, we're doing a lot of work with the interval_map. If it will often say that incremental is fine, why not say it every time? I don't want to look for stalls if there are lots of input sstables.

AFAICT it will say true if there is any disjoint sstable. Maybe we can find a simpler test for it?

**Comment by avikivity** [comment]
> > @denesb please consider merging it.
> 
> I want an ack from @avikivity.

I'm still unhappy about the stall magnet interval map in there.

**Comment by raphaelsc** [comment]
v4:
unconditionally enable incremental for ics / lcs.

File: compaction/compaction.cc (hunk 3 above)
Comment (raphaelsc) [reply compaction/compaction.cc]: An ugly and lazy alternative is to keep contains_multi_fragment_runs, expose a force_incremental_compaction, which off-strategy sets to true. But you're right. I think we should just unconditionally enable incremental for ics / lcs. On the regular path, the cost of enabling incremental on level 0 is low to almost none.
